addrset/addrset.go

```diff
@@ -0,0 +1,106 @@
+package addrset
+
+import (
+	ma "github.com/multiformats/go-multiaddr"
+)
+
+// AddrState describes the state of a dialable addresses as seen by AutoNAT.
+type AddrState int
+
+const (
+	// StateUnknown is used for new entries.
+	StateUnknown AddrState = iota
+	// UnconfirmedListenUnknown applies to interface listen addresses that have unknown reachability.
+	UnconfirmedListenUnknown
+
+	//  UnconfirmedExternalUnknown applies to external observed addresses with unknown reachability.
+	UnconfirmedExternalUnknown
+
+	// Private is applied to addresses that have been deemed to have private reachability by AutoNAT.
+	Private
+
+	// Confirmed is applied to peers that have been confirmed to be dialable by AutoNAT.
+	Confirmed
+)
+
+// DialAddrsSet maintains the state of a dialable addresses as seen by AutoNAT.
+// The state is a set of  multi-addresses, each labeled with a AddrState.
+type DialAddrsSet struct {
+	// all known addresses
+	all []dialAddrState
```

Comment:
I'd consider just using a map keyed on `string(maddr.Bytes())`. The number of allocations should be pretty minimal, relatively.